### PR TITLE
[WIP] Org nonuser signup

### PR DESCRIPTION
**Pull request by camhashemi**
@andrewconner @leogrim @ryanpbrewster @stkem 

the goal of this is to quickly build off of the library non-user invite flow, using the auth token to grant view access to a non-user, sign up through a modal, and auto-join them to the org.

I still need to get the front-end logic completed, but I have a couple backend tests confirming that a user can see an org if they have a valid auth token, and when they sign up they are auto-joined into the org.

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala [hunk 1]
```diff
@@ -263,10 +263,10 @@ trait UserActions extends Logging { self: Controller =>
         userIdOpt match {
           case Some(userId) => buildUserAction(userId, block)
           case None => block(userActionsHelper.buildNonUserRequest).map { resp =>
-            if (resp.header.status == 200 && resp.header.headers.get("Content-Type").exists(_.contains("text/html"))) {
-              maybeAugmentKcid(resp)
-            } else resp
-          }
+              if (resp.header.status == 200 && resp.header.headers.get("Content-Type").exists(_.contains("text/html"))) {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala:266]: I prefer to use `OK` over `200`. easier to understand. ((Similarly, `FORBIDDEN` over `403` and `NOT_FOUND` over `404`, etc.))

Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala:266]: np

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationRequests.scala [hunk 2]
```diff
@@ -74,6 +74,7 @@ object OrganizationFail {
   case object BAD_PARAMETERS extends OrganizationFail(BAD_REQUEST, "bad_parameters")
   case object INVITATION_NOT_FOUND extends OrganizationFail(BAD_REQUEST, "invitation_not_found")
   case object ALREADY_A_MEMBER extends OrganizationFail(BAD_REQUEST, "already_a_member")
+  case object AUTH_INVALID extends OrganizationFail(BAD_REQUEST, "auth_invalid")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationRequests.scala:77]: I think this should be a `UNAUTHORIZED` (i.e., 401).

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationRequests.scala:77]: Oh, this means `INVALID_AUTHTOKEN`? Can we use that instead?

Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationRequests.scala:77]: yup

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala [hunk 3]
```diff
@@ -60,10 +61,17 @@ trait OrganizationAccessActions {
...
               }
-            case Failure(e) => Future.successful(OrganizationFail.INVALID_PUBLIC_ID.asErrorResponse)
+            case request: NonUserRequest[_] =>
+              val authTokenOpt = request.getQueryString("authToken")
+              if (authTokenOpt.exists(authToken => orgInviteCommander.isAuthValid(orgId, authToken))) {
+                val memberPermissions = orgMembershipCommander.getPermissions(orgId, None)
+                block(OrganizationRequest(request, orgId, authTokenOpt, memberPermissions))
+              } else {
+                Future.successful(OrganizationFail.AUTH_INVALID.asErrorResponse)
+              }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala:71]: Good change, I wasn't thinking far enough ahead when I wrote this.

File: kifi-backend/modules/shoebox/test/com/keepit/controllers/website/InviteControllerTest.scala [hunk 4]
```diff
@@ -124,7 +124,7 @@ class InviteControllerTest extends Specification with ShoeboxApplicationInjector
         // README: if you get a 404 here make sure you have pulled the marketing submodule modules/shoebox/marketing
         // 1) run from repository root: git submodule update --init
         // 2) cd to modules/shoebox/marketing and run: npm install
-        // 3) run "mkt" in SBT to generate the file this test depends on (you should only have to do this once)
+        // 3) run "mkt" in SBT from /kifi-backend/ to generate the file this test depends on (you should only have to do this once)
         code === 200
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/InviteControllerTest.scala:128]: I know this isn't yours, but could you change this to `OK` as well?

**Comment by ryanpbrewster** [comment]
This looks awesome. I say merge when ready.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/AuthCommander.scala [hunk 5]
```diff
@@ -470,4 +472,20 @@ class AuthCommander @Inject() (
     }.getOrElse(false)
   }
 
+  def autoJoinOrg(userId: Id[User], orgPubId: PublicId[Organization], authToken: String): Boolean = {
+    Organization.decodePublicId(orgPubId).map { orgId =>
+      implicit val context = HeimdalContext(Map())
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/AuthCommander.scala:477]: I think `HeimdalContext.empty` is a thing.

**Comment by camhashemi** [comment]
merging #16708 (only backend) first to allow me to test in front-end dev mode, then will merge the front-end code

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/AuthCommander.scala [hunk 6]
```diff
@@ -470,4 +472,20 @@ class AuthCommander @Inject() (
...
+          false
+        },
+        { org =>
+          log.info(s"[finishSignup] user(id=$userId) has successfully joined organization $org")
+          true
+        }
+      )
+    }.getOrElse(false)
+  }
+
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/AuthCommander.scala:490]: Not sure what your opinions are regarding `Either.fold` vs an explicit `match`. I think `fold` is super neat, but it has never really struck me as easier to understand.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala [hunk 7]
```diff
@@ -432,4 +432,8 @@ class OrganizationInviteCommanderImpl @Inject() (db: Database,
         suggestedUsers ++ suggestedEmailAddresses
     }
   }
+
+  def isAuthValid(orgId: Id[Organization], authToken: String): Boolean = {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:436]: Let's be explicit and say `AuthToken`.

**Comment by ryanpbrewster** [comment]
Yeah, you're brave for venturing into the frontend code. Good luck!

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/AuthCommander.scala (hunk 6 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/AuthCommander.scala:490]: I think `match` is easier to understand (it's more imperative), but `fold` is more durable/sexy for complex logic, being more functional.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala (hunk 7 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationInviteCommander.scala:436]: I have a consolidation (between library auth and org auth) to do once I confirm that this all works, will add it then.

**Pull request merged**